drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,627 @@
+	struct pinnacle_sample sample;
+	const struct device *dev;
+	struct gpio_callback dr_cb_data;
+	struct k_work work;
+};
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+static bool pinnacle_is_ready_i2c(const struct pinnacle_bus *bus)
+{
+	if (!device_is_ready(bus->i2c.bus)) {
+		LOG_ERR("I2C bus %s is not ready", bus->i2c.bus->name);
+		return false;
+	}
+
+	return true;
+}
+
+static int pinnacle_write_i2c(const struct pinnacle_bus *bus, uint8_t address, uint8_t value)
+{
+	return -1;
+}
+
+static int pinnacle_read_i2c(const struct pinnacle_bus *bus, uint8_t address, uint8_t *value)
+{
+	uint8_t reg = PINNACLE_READ_REG(address);
+
+	int ret = i2c_write_read_dt(&bus->i2c, &reg, 1, value, 1);
+
+	if (ret) {
+		LOG_ERR("Failed to read from I2C bus %s: %d", bus->i2c.bus->name, ret);
+	}
+
+	return 0;
+}
+
+static int pinnacle_seq_read_i2c(const struct pinnacle_bus *bus, uint8_t address, uint8_t *data,
+				 uint8_t count)
+{
+	return -1;
+}
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) */
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+static bool pinnacle_is_ready_spi(const struct pinnacle_bus *bus)
+{
+	if (!spi_is_ready_dt(&bus->spi)) {
+		LOG_ERR("SPI bus %s is not ready", bus->spi.bus->name);
+		return false;
+	}
+
+	return true;
+}
+
+static int pinnacle_write_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t value)
+{
+	uint8_t tx_data[] = {
+		PINNACLE_WRITE_REG(address),
+		value,
+	};
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = ARRAY_SIZE(tx_data),
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	int ret = spi_write_dt(&bus->spi, &tx_set);
+
+	if (ret) {
+		LOG_ERR("Failed to write to SPI %s", bus->spi.bus->name);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int pinnacle_read_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t *value)
+{
+	uint8_t tx_data[] = {
+		PINNACLE_READ_REG(address),
+		PINNACLE_FB,
+		PINNACLE_FB,
+		PINNACLE_FB,
+	};
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = ARRAY_SIZE(tx_data),
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	uint8_t rx_data[4];
+	const struct spi_buf rx_buf[] = {{
+		.buf = rx_data,
+		.len = ARRAY_SIZE(rx_data),
+	}};
+	const struct spi_buf_set rx_set = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	int ret = spi_transceive_dt(&bus->spi, &tx_set, &rx_set);
+
+	if (ret) {
+		LOG_ERR("Failed to read from SPI %s", bus->spi.bus->name);
+		return ret;
+	}
+
+	*value = rx_data[3];
+
+	return 0;
+}
+
+static int pinnacle_seq_read_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t *data,
+				 uint8_t count)
+{
+
+	uint8_t size = count + 3;
+	uint8_t tx_data[size];
+
+	tx_data[0] = PINNACLE_READ_REG(address);
+	tx_data[1] = PINNACLE_FC;
+	tx_data[2] = PINNACLE_FC;
+
+	uint8_t i = 3;
+
+	for (; i < (count + 2); ++i) {
+		tx_data[i] = PINNACLE_FC;
+	}
+
+	tx_data[i++] = PINNACLE_FB;
```

Comment:
could you use some form of `memset(tx_data + 1, PINNACLE_FC, <something>)` here? This is hard to understand